File: src/main/java/me/deecaad/weaponmechanicsplus/weapon/modifiers/attachments/Attachment.java
package me.deecaad.weaponmechanicsplus.weapon.modifiers.attachments;

import me.deecaad.core.file.SerializeData;
import me.deecaad.core.file.SerializerException;
import me.deecaad.core.file.serializers.ItemSerializer;
import me.deecaad.core.mechanics.Mechanics;
import me.deecaad.weaponmechanics.utils.CustomTag;
import me.deecaad.weaponmechanicsplus.WeaponMechanicsPlusAPI;
import me.deecaad.weaponmechanicsplus.weapon.modifiers.ModifierBase;
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.Whitelist;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Attachment extends ModifierBase implements Comparable<Attachment> {

    private String attachmentTitle;
    private int maximumStackAmount;
    private ItemStack item;
    private Set<String> attachmentRequireList;
    private Set<String> attachmentDenyList;
    private Whitelist<String> weaponWhitelist;
    private Mechanics equipMechanics;
    private Mechanics dequipMechanics;

    private Unlockable unlockable;

    /**
     * Default constructor for serializer
     */
    public Attachment() {
    }

    public Attachment(String attachmentTitle, int maximumStackAmount, ItemStack item, Set<String> attachmentRequireList, Set<String> attachmentDenyList,
                      Whitelist<String> weaponWhitelist, Unlockable unlockable, Mechanics equipMechanics, Mechanics dequipMechanics) {
        this.attachmentTitle = attachmentTitle;
        this.maximumStackAmount = maximumStackAmount;
        this.item = item;
        this.attachmentRequireList = attachmentRequireList;
        this.attachmentDenyList = attachmentDenyList;
        this.weaponWhitelist = weaponWhitelist;
        this.unlockable = unlockable;
        this.equipMechanics = equipMechanics;
        this.dequipMechanics = dequipMechanics;
    }

    public String getAttachmentTitle() {
        return attachmentTitle;
    }

    public int getMaximumStackAmount() {
        return maximumStackAmount;
    }

    public ItemStack getItem() {
        return item;
    }

    public Set<String> getAttachmentRequireList() {
        return attachmentRequireList;
    }

    public Set<String> getAttachmentDenyList() {
        return attachmentDenyList;
    }

    public Whitelist<String> getWeaponWhitelist() {
        return weaponWhitelist;
    }

    public Mechanics getEquipMechanics() {
        return equipMechanics;
    }

    public Mechanics getDequipMechanics() {
        return dequipMechanics;
    }

    public Unlockable getUnlockable() {
        return unlockable;
    }

    /**
     * Returns <code>true</code> if at least 1 more of this attachment can be
     * added to the given weapon.
     *
     * @param weapon The non-null weapon to check.
     * @return true if the attachment can be added.
     */
    public boolean canAttach(ItemStack weapon) {
        String weaponTitle = CustomTag.WEAPON_TITLE.getString(weapon);
        if (weaponTitle == null)
            throw new IllegalArgumentException();

        if (!weaponWhitelist.isWhitelisted(weaponTitle))
            return false;

        // If there are no attachments currently on the weapon, then we can attach
        Attachment[] attached = WeaponMechanicsPlusAPI.getAttachments(weapon);
        if (attached == null)
            return true;

        int duplicateCount = 0;
        for (Attachment attachment : attached) {
            if (attachment == this) {
                duplicateCount++;
                continue;
            }

            // Some attachments are not compatible with each other
            if (!attachmentRequireList.isEmpty() && !attachmentRequireList.contains(attachment.attachmentTitle))
                return false;
            if (attachmentDenyList.contains(attachment.attachmentTitle))
                return false;
        }

        // Cannot attach the same attachment multiple times
        return duplicateCount >= maximumStackAmount;
    }

    public void attach(ItemStack weapon) {
        int[] array = CustomTag.ATTACHMENTS.getArray(weapon);
        if (array == null)
            array = new int[0];

        List<Attachment> attachments = new ArrayList<>(array.length + 1);
        for (int id : array) {
            attachments.add(AttachmentRegistry.INSTANCE.get(id));
        }

        // Add the new attachment and sort
        attachments.add(this);
        attachments.sort(null);

        int[] newArray = new int[array.length + 1];
        for (int i = 0; i < attachments.size(); i++) {
            newArray[i] = AttachmentRegistry.INSTANCE.getId(attachments.get(i));
        }

        CustomTag.ATTACHMENTS.setArray(weapon, newArray);
    }

    @Override
    public int compareTo(@NotNull Attachment o) {
        return Integer.compare(priority, o.priority);
    }

    @NotNull
    @Override
    public Attachment serialize(SerializeData data) throws SerializerException {
        String attachmentTitle = data.key;

        int maximumStackAmount = data.of("Maximum_Stack_Amount").assertRange(1, 200).getInt(1);
        ItemStack item = data.of("Item").assertExists().serialize(new ItemSerializer());

        boolean isWeaponWhitelist = data.of("Denying.Weapon_Whitelist").getBool(false);
        List<String> weapons = data.ofList("Denying.Weapons").
                addArgument(String.class, true).assertExists().assertList().get()
                .stream().map(arr -> arr[0]).toList();
        Whitelist<String> weaponWhitelist = new Whitelist<>(isWeaponWhitelist, weapons);

        Set<String> attachmentRequireList = new HashSet<>();
        Set<String> attachmentDenyList = new HashSet<>();
        for (String[] split : data.ofList("Denying.Attachments")
                .addArgument(State.class, true)
                .addArgument(String.class, true)
                .assertList().get()) {

            if (split[0].equalsIgnoreCase("deny"))
                attachmentDenyList.add(split[1]);
            else
                attachmentRequireList.add(split[1]);
        }

        Unlockable unlockable = data.of("Unlockable").serialize(Unlockable.class);

        Mechanics equipMechanics = data.of("Equip_Mechanics").serialize(Mechanics.class);
        Mechanics dequipMechanics = data.of("Dequip_Mechanics").serialize(Mechanics.class);

        ModifierBase base = super.serialize(data);
        Attachment returnValue = new Attachment(attachmentTitle, maximumStackAmount, item, attachmentRequireList, attachmentDenyList, weaponWhitelist, unlockable, equipMechanics, dequipMechanics);
        returnValue.priority = base.getPriority();
        returnValue.modifier = base.getModifier();
        returnValue.perWeaponModifiers = base.getPerWeaponModifiers();
        return returnValue;
    }

    private enum State {
        DENY, REQUIRE
    }
}

File: src/main/kotlin/me/deecaad/weaponmechanicsplus/weapon/modifiers/attachments/Attachment.kt
package me.deecaad.weaponmechanicsplus.weapon.modifiers.attachments

import me.deecaad.core.file.*
import me.deecaad.core.file.serializers.ItemSerializer
import me.deecaad.core.mechanics.Mechanics
import me.deecaad.weaponmechanics.utils.CustomTag
import me.deecaad.weaponmechanicsplus.WeaponMechanicsPlusAPI
import me.deecaad.weaponmechanicsplus.weapon.modifiers.ModifierBase
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.Whitelist
import org.bukkit.inventory.ItemStack

class Attachment : ModifierBase, Comparable<Attachment> {

    lateinit var attachmentTitle: String
    var maximumStackAmount = 1
    lateinit var item: ItemStack
    var attachmentRequireList: Set<String> = setOf()
    var attachmentDenyList: Set<String> = setOf()
    lateinit var weaponWhitelist: Whitelist<String>
    var denyMechanics: Mechanics? = null
    var equipMechanics: Mechanics? = null
    var dequipMechanics: Mechanics? = null
    var unlockable: Unlockable? = null

    /**
     * Default constructor for serializer
     */
    constructor()

    constructor(attachmentTitle: String, maximumStackAmount: Int, item: ItemStack, attachmentRequireList: Set<String>,
                attachmentDenyList: Set<String>, weaponWhitelist: Whitelist<String>, unlockable: Unlockable?,
                denyMechanics: Mechanics?, equipMechanics: Mechanics?, dequipMechanics: Mechanics?) {
        this.attachmentTitle = attachmentTitle
        this.maximumStackAmount = maximumStackAmount
        this.item = item
        this.attachmentRequireList = attachmentRequireList
        this.attachmentDenyList = attachmentDenyList
        this.weaponWhitelist = weaponWhitelist
        this.unlockable = unlockable
        this.denyMechanics = denyMechanics
        this.equipMechanics = equipMechanics
        this.dequipMechanics = dequipMechanics
    }

    /**
     * Returns a copy of the attachment item.
     */
    fun generateItem() = item.clone()

    /**
     * Returns `true` if at least 1 more of this attachment can be
     * added to the given weapon.
     *
     * @param weapon The non-null weapon to check.
     * @return true if the attachment can be added.
     */
    fun canAttach(weapon: ItemStack?): Boolean {
        val weaponTitle = CustomTag.WEAPON_TITLE.getString(weapon) ?: throw IllegalArgumentException()
        if (!weaponWhitelist.isWhitelisted(weaponTitle)) return false

        // If there are no attachments currently on the weapon, then we can attach
        val attached = WeaponMechanicsPlusAPI.getAttachments(weapon) ?: return true
        var duplicateCount = 0
        for (attachment in attached) {
            if (attachment === this) {
                duplicateCount++
                continue
            }

            // Some attachments are not compatible with each other
            if (attachmentRequireList.isNotEmpty() && !attachmentRequireList.contains(attachment.attachmentTitle)) return false
            if (attachmentDenyList.contains(attachment.attachmentTitle)) return false
        }

        // Cannot attach the same attachment multiple times
        return duplicateCount < maximumStackAmount
    }

    fun attach(weapon: ItemStack?) {
        var array = CustomTag.ATTACHMENTS.getStringArray(weapon)
        if (array == null)
            array = emptyArray()

        val attachments: MutableList<Attachment> = ArrayList(array.size + 1)
        for (id in array) {
            val attachment = AttachmentRegistry.INSTANCE[id]

            // attachment is null when the admin deletes an attachment from
            // config after it is attached to a weapon.
            if (attachment != null)
                attachments.add(attachment)
        }

        // Add the new attachment and sort
        attachments.add(this)
        attachments.sort()
        CustomTag.ATTACHMENTS.setStringArray(weapon, attachments.map { it.attachmentTitle }.toTypedArray())
    }

    override fun compareTo(other: Attachment): Int {
        return priority.compareTo(other.priority)
    }

    @Throws(SerializerException::class)
    override fun serialize(data: SerializeData): Attachment {
        val attachmentTitle = data.key
        val maximumStackAmount = data.of("Maximum_Stack_Amount").assertRange(1, 100).getInt(1)

        // Make sure to use CustomTag during serialization so recipe works
        data.of("Item").assertExists()
        val tags = mapOf(Pair(CustomTag.ATTACHMENT_TITLE.key, attachmentTitle))
        val item = ItemSerializer().serializeWithTags(data.move("Item"), tags)

        val isWeaponWhitelist = data.of("Denying.Weapon_Whitelist").getBool(false)
        val weapons = data.ofList("Denying.Weapons").addArgument(String::class.java, true).assertList().get()
                .stream().map { arr: Array<String> -> arr[0] }.toList()
        val weaponWhitelist = Whitelist(isWeaponWhitelist, weapons)

        val attachmentRequireList: MutableSet<String> = HashSet()
        val attachmentDenyList: MutableSet<String> = HashSet()
        for (split in data.ofList("Denying.Attachments")
            .addArgument(State::class.java, true)
            .addArgument(String::class.java, true)
            .assertList().get()) {

            if (split[0].equals("deny", ignoreCase = true))
                attachmentDenyList.add(split[1])
            else
                attachmentRequireList.add(split[1])
        }

        val unlockable = data.of("Unlockable").serialize(Unlockable::class.java)
        val denyMechanics = data.of("Denying.Mechanics").serialize(Mechanics::class.java)
        val equipMechanics = data.of("Attach_Mechanics").serialize(Mechanics::class.java)
        val dequipMechanics = data.of("Detach_Mechanics").serialize(Mechanics::class.java)

        val returnValue = Attachment(attachmentTitle, maximumStackAmount, item, attachmentRequireList, attachmentDenyList, weaponWhitelist, unlockable, denyMechanics, equipMechanics, dequipMechanics)

        val base = super.serialize(data)
        returnValue.priority = base.priority
        returnValue.modifier = base.modifier
        returnValue.perWeaponModifiers = base.perWeaponModifiers
        return returnValue
    }

    private enum class State {
        DENY, REQUIRE
    }
}